File: 18. MissingNumber_XOR.py
# Topic: Find missing number in array containing 1..N using XOR

def missing_number(arr, n):
    xr = 0
    for i in range(1, n+1): xr ^= i
    for x in arr: xr ^= x
    return xr

if __name__ == '__main__':
    arr = [1,2,4,5,6]
    print('Missing:', missing_number(arr,6))

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 18. MissingNumber_XOR.cpp
// Topic: Find missing number in array containing 1..N using XOR
#include <bits/stdc++.h>
using namespace std;

int missingNumber(vector<int>& arr, int n)
{
    int xr = 0;
    for (int i = 1; i <= n; ++i) xr ^= i;
    for (int x : arr) xr ^= x;
    return xr;
}

int main()
{
    vector<int> arr = {1, 2, 4, 5, 6};
    cout << "Missing: " << missingNumber(arr, 6) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)
